pkg/roll/run_update.go

```diff
@@ -152,24 +152,32 @@ ROLL_LOOP:
 				break
 			}
 
-			if newNodes.Desired >= u.DesiredReplicas && newNodes.Healthy >= u.MinimumReplicas {
-				// We only ask for u.MinimumReplicas nodes to be healthy
-				// before declaring an upgrade to be complete.
-				// This is so that if a deployer intentionally deploys a known-bad SHA
-				// for which no nodes become healthy (specifying minimum == 0),
-				// we don't leave an RU and old RC lying around.
-				// The RU and RC would have required manual intervention to clean up.
-				//
+			if newNodes.Desired >= u.DesiredReplicas {
 				// The above condition contains a >= instead of a ==.
 				// This is despite the fact that normally rollAlgorithm will never cause
 				// newDesired > u.Desired.
 				// The benefit of using >= is that it allows termination even if Run
 				// resumes from an unusual state.
-				u.logger.WithFields(logrus.Fields{
-					"old": oldNodes,
-					"new": newNodes,
-				}).Debugln("Upgrade complete")
-				break ROLL_LOOP
+
+				if newNodes.Healthy >= u.MinimumReplicas {
+					// We only ask for u.MinimumReplicas nodes to be healthy
+					// before declaring an upgrade to be complete.
+					// This is so that if a deployer intentionally deploys a known-bad SHA
+					// for which no nodes become healthy (specifying minimum == 0),
+					// we don't leave an RU and old RC lying around.
+					// The RU and RC would have required manual intervention to clean up.
+					u.logger.WithFields(logrus.Fields{
+						"old": oldNodes,
+						"new": newNodes,
+					}).Debugln("Upgrade complete")
+					break ROLL_LOOP
+				} else {
```

Comment:
My understanding of the issue you describe is if I have nodes A B C D E, and ask for a canary of 1 node, and A is chosen but never becomes healthy, the deploy blocks forever.

I think it has to. The alternative is to keep trying to bring node B in. What if B becomes unhealthy too? We wouldn't want to bring C, D, E in to have them all become unhealthy (granted, because of minimum health guarantees this should only happen if the deployer set minimum health to 0, which we hope doesn't happen)